### ISSUE TITLE
MaskedTable, indices & join

### ISSUE THREAD

**Issue by abeelen** [astropy/astropy#9289]
When working on MaskedTable, I have to remove indices before joining Tables if I want to be able to use indices on the join MaskedTable :

```python
from astropy.table import Table, join
t = Table([('w', 'x', 'y', 'z'), (10, 1, 9, 8)], names=('a', 'b'), dtype=['str', 'i8'], masked=True)
t.add_index('a')
v = Table([('a', 'b', 'c', 'd'), (10, 1, 9, 9)], names=('c', 'b'), dtype=['str', 'i8'])
w = join(t, v, keys='b', join_type='outer')
w.add_index('a')
print(w.loc['x'])
```
will produce a `TypeError: list indices must be integers or slices, not NoneType`. The error is not present if : 
* the index is removed before the join `t.remove_indice('a')` 
* the table is not masked

Could this related to the fact that the joined MaskedTable keep its indices, while the joined Table loose it ?

Tested with astropy 3.2.1 and python 3.7.4

**Comment by pllim** [comment]
@abeelen , I cannot reproduce the error with `astropy` 4.0.dev25735. Can you please try with the latest `master` version and see? I get:
```
>>> w.loc['x']                                                             
<Row index=0>
 a     b    c  
str1 int64 str1
---- ----- ----
   x     1    b
```

**Comment by bsipocz** [comment]
There has been big overhauls around masks for Tables, and this indeed has been fixed on dev (while I see the issue on 3.2.x).

**Comment by abeelen** [comment]
Thanks all for having a look at this. I will wait for the 4.x release then, in the meantime removing the indices or the masks and reapplying them would work
